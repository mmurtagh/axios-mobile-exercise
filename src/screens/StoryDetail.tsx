import React, { useContext } from 'react';
import { NativeStackScreenProps } from '@react-navigation/native-stack';
import { observer } from 'mobx-react-lite';
import styled from 'styled-components/native';
import { Linking } from 'react-native'

import { RootStackParamList, StoryStoreContext } from '../index';
import { Button, Headline, Screen, Card, Image, Paragraph } from '../components/components';
import { DetailCaption } from '../components/DetailCaption';
import { BlockText } from '../components/BlockText';
import { fontColor, backdropColor, spacing } from '../utils/styling';
import { ImageSourcePropType } from 'react-native';

export type Props = NativeStackScreenProps<RootStackParamList, 'StoryDetail'>;

const axiosUrl = 'https://www.axios.com/'

const HeadlineCard = styled(Card)`
  margin-top: ${spacing()}
`
const ScrollView = styled.ScrollView`
  background-color: ${backdropColor};
  height: 100%;
`

const Break = styled.View`
  border-bottom-width: 1px;
  border-color: ${fontColor}
  margin-vertical: ${spacing('lg')}
`;

const BlockTextCard = styled(Card)`
  margin-vertical: ${spacing()};
`;

const StoryHeadline = styled(Headline)`
  text-align: center;
  font-weight: bold;
`

/** 
 * @name: StoryDetail
 * @description: Screen displaying the detail view of a story
 * @param story: The story the list item represents
 * @param index: Where the story falls in the list
 * @param totalStories: The total number of stories in the list
 * @param id: The id of the story
*/
export const StoryDetail = observer(({ route: { params: { id } } }: Props) => {
  const store = useContext(StoryStoreContext);

  const story = store.getStory(id);

  if (story === null) return null

  const imageDescription = story.imageDescription;
  const imageSources: Crop[] = story.getImageSources('16x9');

  return (
    <Screen>
      <ScrollView>
        <HeadlineCard>
          <StoryHeadline
            accessibilityRole="header"
            accessibilityLabel={story.headline}
            testID="headline"
          >
            {story.headline}
          </StoryHeadline>
          <Break />
          <Image
            accessible
            accessibilityRole="image"
            accessibilityLabel={imageDescription}
            testID="image"
            aspectRatio={16 / 9}
            source={imageSources}
          />
          <DetailCaption story={story} />
        </HeadlineCard>
        <BlockTextCard testID="block-text-card">
          {story.blocks.map((block) => {
            return <BlockText key={block.key} block={block} />
          })}
          <Button
            testID="button"
            icon="open-in-new"
            title="Visit Axios.com"
            onPress={() => Linking.openURL(axiosUrl)}
          />
        </BlockTextCard>
      </ScrollView>
    </Screen>
  );
});

export default StoryDetail;
