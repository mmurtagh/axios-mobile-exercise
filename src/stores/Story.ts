import { makeAutoObservable, toJS } from 'mobx';
import dayjs from 'dayjs';

export enum SupportedEntityType {
  LINK = 'LINK',
}

export enum SupportedBlockType {
  UNSTYLED = 'unstyled',
  UNORDERED_LIST_ITEM = 'unordered-list-item',
}

export enum SupportedInlineStyle {
  BOLD = 'BOLD',
}

/** 
 * @name sanitizeContentInstance
 * @description: Filters out unsupported features from raw AxiosContentInstances.
 * @param _contentInstance: The AxiosContentInstance object that needs to be sanitized.
*/
function sanitizeContentInstance (_contentInstance: AxiosContentInstance): AxiosContentInstance {
  const contentInstance = { ..._contentInstance }
  const { blocks: { blocks } } = contentInstance;

  // Filter out all blocks with unsupported block type
  contentInstance.blocks.blocks = blocks.filter(({ type }) => {
    const supportedTypes: string[] = Object.values(SupportedBlockType);

    return supportedTypes.includes(type)
  })

  return contentInstance
}

/** 
 * @class: Story
 * @description: An Axios story. A mobx domain object managed by StoryStore.js.
*/
export class Story {
  contentInstance: AxiosContentInstance;

  constructor(contentInstance: AxiosContentInstance) {
    this.contentInstance = sanitizeContentInstance(contentInstance);
    makeAutoObservable(this);
  }

  get headline(): string {
    return this.contentInstance.headline;
  }

  get publishedDate(): dayjs.Dayjs {
    return dayjs(this.contentInstance.published_date);
  }

  /** 
   * @name getImageSources
   * @description: Gets the list of image sources of a given ratio.
   * @param ratio: The desired image's aspect ratio
  */
   getImageSources (ratio: '1x1' | '4x3' | '16x9'): Crop[] {
    let image = this.contentInstance.primary_image

    if (image === null || !image.crops[ratio].sizes) {
      image = this.contentInstance.social_image;
    }

    if (image === null) {
      return []
    }

    return toJS(image.crops[ratio].sizes)
  }

  get id(): string {
    return this.contentInstance.id;
  }

  get author(): string | null {
    if (!this.contentInstance.authors.length) {
      return null;
    }

    return this.contentInstance.authors[0].display_name;
  }

  get primaryTopicName(): string | null {
    const topics = this.contentInstance.topics;

    if (topics === null || topics.length === 0) {
      return null;
    }

    return topics[0].name;
  }

  get blocks(): Block[] {
    return this.contentInstance.blocks.blocks;
  }

  get entityMap(): Entity[] {
    return this.contentInstance.blocks.entityMap;
  }

  /** 
   * @name imageDescription
   * @description: Gets a description of the image associated with the story.
  */
  get imageDescription(): string {
    let image = this.contentInstance.primary_image

    if (image === null) {
      image = this.contentInstance.social_image;
    }

    if (image === null) {
      return ''
    }

    const { blocks } = image.caption
    
    if (blocks.length === 0) {
      return ''
    }

    return blocks[0].text;
  }
}
