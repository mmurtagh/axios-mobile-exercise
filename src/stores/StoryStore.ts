import { makeAutoObservable, runInAction } from 'mobx';

import { AxiosService } from '../services/AxiosService';
import { Story } from './Story';

/** 
 * @class: StoryStore
 * @description: A mobx domain store representing a store of Axios stories.
*/
export class StoryStore {
  service: AxiosService;
  isLoadingStories = false;
  stories: Story[] = [];

  constructor() {
    makeAutoObservable(this);
    this.service = new AxiosService();
  }

  /** 
   * @class: loadMostRecentStories
   * @description: Loads the most recent stories using the Axios stream and content APIs.
  */
  async loadMostRecentStories(): Promise<void> {
    const stream: AxiosStream = await this.service.getStream();

    const contentCalls = stream.results.map(id => {
      return this.service.getContent(id);
    });

    const data: AxiosContentInstance[] = await Promise.all(contentCalls);

    runInAction(() => {
      this.stories = data.map((story) => new Story(story))
    });
  }

  getStory(id: string): Story | null {
    return (
      this.stories.find(story => {
        return id === story.id;
      }) || null
    );
  }
}
