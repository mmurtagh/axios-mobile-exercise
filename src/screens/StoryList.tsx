import React, { useContext, useEffect, useState } from 'react';
import { observer } from 'mobx-react-lite';
import styled from 'styled-components/native';
import { FlatList } from 'react-native';

import { Screen } from '../components/components';
import { Story } from '../stores/Story';
import { StoryListItem } from '../components/StoryListItem';
import { spacing } from '../utils/styling';
import { StoryStore } from '../stores/StoryStore';
import { StoryStoreContext } from '..';
import { ErrorCard } from '../components/ErrorCard';

const Separator = styled.View`
  padding-top: ${spacing('sm')};
`
const Header = styled.View`
  padding-top: ${spacing()};
`
const Spinner = styled.ActivityIndicator`
  padding-top: ${spacing()};
`
const RefreshControl = styled.RefreshControl``
enum LoadingState {
  LOADING,
  READY,
  REFRESHING,
  ERRORED,
}

/** 
 * @name: StoryList
 * @description: Screen displaying the list of stories
*/
export const StoryList = observer(() => {
  const store: StoryStore = useContext(StoryStoreContext);
  const [ loadingState, setLoadingState ] = useState(LoadingState.LOADING)

  const loadStories = (state: LoadingState) => {
    setLoadingState(state);

    store.loadMostRecentStories()
      .then(() => setLoadingState(LoadingState.READY))
      .catch(() => setLoadingState(LoadingState.ERRORED))
  }

  useEffect(() => {
    loadStories(LoadingState.LOADING); 
  }, []);

  const renderItem = ({ item, index }: { item: Story, index: number }) => {
    return <StoryListItem story={item} index={index} totalStories={store.stories.length} />;
  };

  return (
    <Screen>
      {loadingState === LoadingState.LOADING && <Spinner testID="spinner" size="large"/>}
      {loadingState === LoadingState.ERRORED && <ErrorCard onPress={() => loadStories(LoadingState.LOADING)}/>}
      {loadingState !== LoadingState.ERRORED && (
        <FlatList
          testID="flatlist"
          ItemSeparatorComponent={Separator}
          data={loadingState === LoadingState.LOADING ? [] : store.stories}
          renderItem={renderItem}
          ListHeaderComponent={Header}
          ListFooterComponent={Header}
          refreshControl={
            <RefreshControl
              testID="refresh-control"
              refreshing={loadingState === LoadingState.REFRESHING}
              onRefresh={() => loadStories(LoadingState.REFRESHING)}
            />
          }
        />
      )}
    </Screen>
  );
});
