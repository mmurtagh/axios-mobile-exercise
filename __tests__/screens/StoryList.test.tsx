import React from 'react';
import { act, fireEvent, render, waitFor } from '@testing-library/react-native';
import { clearTimers } from 'mobx-react-lite';

import { StoryList } from '../../src/screens/StoryList'
import { StoryStore } from '../../src/stores/StoryStore';

const mockStore = new StoryStore();
mockStore.loadMostRecentStories = jest.fn().mockResolvedValue(null);

jest.mock('react', () => {
  return {
    ...jest.requireActual('react'),
    useContext: () => mockStore,
  }
})

describe('StoryList', () => {
  beforeEach(() => {
    (mockStore.loadMostRecentStories as jest.Mock).mockClear();
  })

  afterEach(() => {
    clearTimers();
  })

  test('snapshot', async () => {
    const obj = await waitFor(() => render(<StoryList />));

    expect(obj).toMatchSnapshot();
  });

  test('load stores on init', async () => {
    await waitFor(() => render(<StoryList />));

    expect(mockStore.loadMostRecentStories).toBeCalled();
  })

  test('refresh', async () => {
    const { getByTestId } = await waitFor(() => render(<StoryList />));

    (mockStore.loadMostRecentStories as jest.Mock).mockClear();

    const refreshControl = getByTestId('flatlist').props.refreshControl;
    await act(async () => await fireEvent(refreshControl, 'refresh'));

    expect(mockStore.loadMostRecentStories).toBeCalled();
  })

  test('load succeeds', async () => {
    (mockStore.loadMostRecentStories as jest.Mock).mockResolvedValue(null)
    const obj = await waitFor(() => render(<StoryList />));

    expect(mockStore.loadMostRecentStories).toBeCalled();

    expect(obj.queryByTestId('error-card')).toBeFalsy();
  })

  test('load fails', async () => {
    (mockStore.loadMostRecentStories as jest.Mock).mockRejectedValue(null)
    const obj = await waitFor(() => render(<StoryList />));

    expect(mockStore.loadMostRecentStories).toBeCalled();

    expect(obj.queryByTestId('error-card')).toBeTruthy();
  })

  test('load fails then refresh', async () => {
    (mockStore.loadMostRecentStories as jest.Mock).mockRejectedValue(null)
    const obj = await waitFor(() => render(<StoryList />));

    expect(mockStore.loadMostRecentStories).toBeCalled();
    expect(obj.queryByTestId('error-card')).toBeTruthy();

    (mockStore.loadMostRecentStories as jest.Mock).mockResolvedValue(null)

    await act(async () => await fireEvent(obj.getByTestId('error-button'), 'press'))

    expect(mockStore.loadMostRecentStories).toBeCalled();
    expect(obj.queryByTestId('error-card')).toBeFalsy();
  })
});
