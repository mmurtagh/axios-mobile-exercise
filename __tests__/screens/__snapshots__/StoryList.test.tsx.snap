// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryList snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "backgroundColor": "#d3d3d3",
        "flexGrow": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "paddingHorizontal": 10,
        },
      ]
    }
  >
    <RCTScrollView
      ItemSeparatorComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "inlineStyle": InlineStyle {
            "rules": Array [
              "
  padding-top: ",
              "5px",
              ";
",
            ],
          },
          "render": [Function],
          "shouldForwardProp": undefined,
          "styledComponentId": "StyledNativeComponent",
          "target": [Function],
          "withComponent": [Function],
        }
      }
      ListFooterComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "inlineStyle": InlineStyle {
            "rules": Array [
              "
  padding-top: ",
              "10px",
              ";
",
            ],
          },
          "render": [Function],
          "shouldForwardProp": undefined,
          "styledComponentId": "StyledNativeComponent",
          "target": [Function],
          "withComponent": [Function],
        }
      }
      ListHeaderComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "inlineStyle": InlineStyle {
            "rules": Array [
              "
  padding-top: ",
              "10px",
              ";
",
            ],
          },
          "render": [Function],
          "shouldForwardProp": undefined,
          "styledComponentId": "StyledNativeComponent",
          "target": [Function],
          "withComponent": [Function],
        }
      }
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <Styled(RefreshControlMock)
          onRefresh={[Function]}
          refreshing={false}
          testID="refresh-control"
        />
      }
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      testID="flatlist"
      viewabilityConfigCallbackPairs={Array []}
    >
      <RCTRefreshControl />
      <View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Array [
                Object {
                  "paddingTop": 10,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Array [
                Object {
                  "paddingTop": 10,
                },
              ]
            }
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
